### ISSUE TITLE
NVFP4 Selective Fwd / Bwd

### ISSUE THREAD

**Issue by jeromeku** [NVIDIA/TransformerEngine#2569]
The [NVFP4 Pretraining paper](https://arxiv.org/pdf/2509.25149) observed that in cases where there is a loss gap between `NVFP4` and a higher precision baseline that the gap can be closed by switching to higher precision during the last ~10% of training (See Appendix D).

<img width="683" height="412" alt="Image" src="https://github.com/user-attachments/assets/3125e80a-b145-4a00-8b13-ebff5e5ebff4" />

Moreover, they observe that the majority of quantization error results from the forward pass.

What is the recommended way to selectively switch to higher precision *only* for `FProp` and while using `NVFP4` only for backward?

**Comment by zhongbozhu** [comment]
I think by default such options are not provided out of the box. To tweak TE for such options shouldn't be too hard, we just need to make sure we skip the quantization's properly (no new kernels), and don't discard that high precision copy and make sure it's saved for the backward.  The python to C++ bindings are designed to be flexible enough, such that they can take whatever types of inputs, checkout [tex.general_grouped_gemm](https://github.com/NVIDIA/TransformerEngine/blob/69636a08171d162a223ae3a01e4b36902b64dba1/transformer_engine/pytorch/csrc/extensions/gemm.cpp#L403). The gemm doesn't care about recipe, it just checks the type of inputs (fp8, fp4, or non-quantized bf16) and dispatch to the right gemm kernel.
